From 224f0c969a4a69067502b7ed649746188daad2c9 Mon Sep 17 00:00:00 2001
From: Oliver Eftevaag <oliver.eftevaag@qt.io>
Date: Thu, 4 Sep 2025 11:17:34 +0200
Subject: [PATCH 092/239] Menu: prioritize parentItem's window as transient
 parent for native menus

QPlatformMenu::showPopup() takes a QWindow ptr as its first parameter,
which may or may not be used, depending on the platform. But usually
this becomes the 'transient' parent window for the menu.

This QWindow ptr is chosen as the first element in the list returned
from QGuiApplication::topLevelWindows(), and there's no guarantee that
it won't be empty. On macOS, the following snippet would reproduce a
crash, when opening the menu and pressing Command+Q:
```
Button {
    text: "Open menu"
    onClicked: contextMenu.open()
    Menu {
        id: contextMenu
        popupType: Popup.Native
        title: "File"
        MenuItem {
            action: Action {
                text: "Open"
            }
        }
    }
}
```
This would be the backtrace:
```
4  qAbort()
5  void qt_maybe_message_fatal<QString&>(QtMsgType, QMessageLogContext const&, QString&)
6  qt_message(QtMsgType, QMessageLogContext const&, const char *, char *)
7  QMessageLogger::fatal(const char *, ...) const
8  qt_assert(const char *, const char *, int)
9  QList<QWindow *>::first()
10 QQuickMenuPrivate::setNativeMenuVisible(bool)
11 QQuickMenu::setVisible(bool)
12 QQuickPopup::close()
13 QQuickPopup::setParentItem(QQuickItem *)
14 QQuickPopupPrivate::itemDestroyed(QQuickItem *)
15 QQuickMenuPrivate::itemDestroyed(QQuickItem *)
16 void QQuickItemPrivate::notifyChangeListeners<void (QQuickItemChangeListener:: *)(QQuickItem *), QQuickItem *>(QFlags<QQuickItemPrivate::ChangeType>, void (QQuickItemChangeListener:: *&&)(QQuickItem *), QQuickItem *&&)
17 QQuickItem::~QQuickItem()
18 QQuickControl::~QQuickControl()
19 QQuickAbstractButton::~QQuickAbstractButton()
20 QQuickButton::~QQuickButton()
21 QQmlPrivate::QQmlElement<QQuickButton>::~QQmlElement()
22 QQmlPrivate::QQmlElement<QQuickButton>::~QQmlElement()
23 QQmlPrivate::QQmlElement<QQuickButton>::~QQmlElement()
24 QObjectPrivate::deleteChildren()
25 QObject::~QObject()
26 QWindow::~QWindow()
27 QQuickWindow::~QQuickWindow()
28 QQuickWindowQmlImpl::~QQuickWindowQmlImpl()
29 QQuickApplicationWindow::~QQuickApplicationWindow()
30 QQmlPrivate::QQmlElement<QQuickApplicationWindow>::~QQmlElement()
31 QQmlPrivate::QQmlElement<QQuickApplicationWindow>::~QQmlElement()
32 QQmlPrivate::QQmlElement<QQuickApplicationWindow>::~QQmlElement()
33 void qDeleteAll<QList<QObject *>::const_iterator>(QList<QObject *>::const_iterator, QList<QObject *>::const_iterator)
34 void qDeleteAll<QList<QObject *>>(QList<QObject *> const&)
35 QQmlApplicationEnginePrivate::cleanUp()
36 QQmlApplicationEngine::~QQmlApplicationEngine()
37 QQmlApplicationEngine::~QQmlApplicationEngine()
38 main
```

Since we only need this QWindow ptr to open a native menu, there is no
reason to even check that list in cases were we're closing a native
menu. Thus, it makes sense to completely skip that part when closing.

Not only can we skip this step when closing a native menu, but we also
don't really need to check qGuiApp->topLevelWindows() in the first
place, since the window we're interested in is referenced by the parent
item. Using the parent item's window is more semantically correct, since
getting an arbitrary top level window from qGuiApp->topLevelWindows() could
cause QPlatformMenu::showPopup() to get called with a pointer a
completely different window than the intended one.
QPlatformMenu::showPopup() and QHighDpi::toNativeLocalPosition() both
handle cases where the window argument is nullptr.

An autotest would be very welcome, but due to native menus blocking
nature, it doesn't seem feasible.

Pick-to: 6.9 6.8
Change-Id: I7f61ed245307a821fdf567e65e3739566505ba73
Reviewed-by: Mitch Curtis <mitch.curtis@qt.io>
(cherry picked from commit 7ca76d3799b4081f64c4dc2877b52dcc4d03e828)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quicktemplates/qquickmenu.cpp | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/quicktemplates/qquickmenu.cpp b/src/quicktemplates/qquickmenu.cpp
index cf2dce364b..903f7aba55 100644
--- a/src/quicktemplates/qquickmenu.cpp
+++ b/src/quicktemplates/qquickmenu.cpp
@@ -529,14 +529,16 @@ void QQuickMenuPrivate::setNativeMenuVisible(bool visible)
     this->visible = visible;
     syncWithNativeMenu();
 
-    QPoint offset;
-    QWindow *window = effectiveWindow(qGuiApp->topLevelWindows().first(), &offset);
-
     if (visible) {
-        lastDevicePixelRatio = window->devicePixelRatio();
+        QPoint offset;
+        QWindow *window = nullptr;
+        if (parentItem)
+            window = effectiveWindow(parentItem->window(), &offset);
+
+        lastDevicePixelRatio = window ? window->devicePixelRatio() : qGuiApp->devicePixelRatio();
 
         const QPointF globalPos = parentItem->mapToGlobal(x, y);
-        const QPoint windowPos = window->mapFromGlobal(globalPos.toPoint());
+        const QPoint windowPos = window ? window->mapFromGlobal(globalPos.toPoint()) : parentItem->mapToScene(QPoint(x, y)).toPoint();
         QRect targetRect(windowPos, QSize(0, 0));
         auto *daPriv = QQuickItemPrivate::get(parentItem)->deliveryAgentPrivate();
         Q_ASSERT(daPriv);
-- 
2.51.2

